currencyrate/currencyrate.py

```diff
@@ -0,0 +1,148 @@
+    Source('coinbase',
+           'https://api.coinbase.com/v2/prices/spot?currency={currency}',
+           ['data', 'amount']),
+    # e.g. {  "USD" : {"15m" : 6650.3, "last" : 6650.3, "buy" : 6650.3, "sell" : 6650.3, "symbol" : "$"},  "AUD" : {"15m" : 10857.19, "last" : 10857.19, "buy" : 10857.19, "sell" : 10857.19, "symbol" : "$"},...
+    Source('blockchain.info',
+           'https://blockchain.info/ticker',
+           ['{currency}', 'last']),
+]
+
+
+def get_currencyrate(plugin, currency, req_template, response_members):
+    # NOTE: Bitstamp has a DNS/Proxy issues that can return 404
+    # Workaround: retry up to 5 times with a delay
+    currency_lc = currency.lower()
+    url = req_template.format(currency_lc=currency_lc, currency=currency)
+    for _ in range(5):
+        r = requests.get(url, proxies=plugin.proxies)
+        if r.status_code != 200:
+            time.sleep(1)
+            continue
+        break
+
+    if r.status_code != 200:
+        plugin.log(level='info', message='{}: bad response {}'.format(url, r.status_code))
+        return None
+
+    json = r.json()
+    for m in response_members:
+        expanded = m.format(currency_lc=currency_lc, currency=currency)
+        if expanded not in json:
+            plugin.log(level='debug', message='{}: {} not in {}'.format(url, expanded, json))
+            return None
+        json = json[expanded]
+
+    try:
+        return Millisatoshi(int(10**11 / float(json)))
+    except Exception:
+        plugin.log(level='info', message='{}: could not convert {} to msat'.format(url, json))
+        return None
+
+
+def set_proxies(plugin):
+    config = plugin.rpc.listconfigs()
+    if 'always-use-proxy' in config and config['always-use-proxy']:
+        paddr = config['proxy']
+        # Default port in 9050
+        if ':' not in paddr:
+            paddr += ':9050'
+        plugin.proxies = {'https': 'socks5h://' + paddr,
+                          'http': 'socks5h://' + paddr}
+    else:
+        plugin.proxies = None
+
+
+# Don't grab these more than once per hour.                        
+@cached(cache=TTLCache(maxsize=1024, ttl=3600))
+def get_rates(plugin, currency):
+    rates = {}
+    for s in sources:
+        r = get_currencyrate(plugin, currency, s.urlformat, s.replymembers)
+        if r is not None:
+            rates[s.name] = r
+
+    return rates
+
+
+@plugin.method("currencyrate")
+def currencyrate(plugin, currency):
+    """Gets currency from given APIs."""
+
+    return get_rates(plugin, currency.upper())
+
+
+@plugin.method("currencyconvert")
+def currencyconvert(plugin, amount, currency):
+    """Converts currency using given APIs."""
+    rates = get_rates(plugin, currency.upper())
+    if len(rates) == 0:
+        raise Exception("No values available for currency {}".format(currency.upper()))
+    val = statistics.median([m.millisatoshis for m in rates.values()]) * float(amount)
+    return {"msat": Millisatoshi(round(val))}
```

Comment:
Do we want to notify or warn about if `len(rates)` is less than certain thresholds?
 - `< len(sources)` # less than defined sources
 - `< 3`  # a too small number
 - ... ?

Reply:
I'd rather have some warning when it's markedly different from previous values, TBH (implying some persistence).  Though with Bitcoin's price, it's hard to know what is a bug, and what is a feature!